classes/task/delay_delete_external_objects.php

```diff
@@ -0,0 +1,79 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Task that checks for old orphaned objects, and removes their metadata (record)
+ * and external file (if delete external enabled) as it is no longer useful/relevant.
+ *
+ * @package   tool_objectfs
+ * @author    Kevin Pham <kevinpham@catalyst-au.net>
+ * @copyright Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_objectfs\task;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once(__DIR__ . '/../../lib.php');
+
+class delay_delete_external_objects extends task {
+
+    /** @var string $stringname */
+    protected $stringname = 'delay_delete_external_objects_task';
+
+    /**
+     * Execute task
+     */
+    public function execute() {
+        global $DB;
+
+        $timeperiodforremoval = $this->config->delaydeleteexternalobject;
+        if (empty($timeperiodforremoval)) {
+            mtrace('Skipping delayed deletion of the external object of the delaydeleteexternalobject is set to an empty value.');
```

Comment:
What happens if there are files in the table still? Should we purge the table then?

Reply:
Right now the records are not deleting from the table, only updating. We could delete those directly from the table instead.

Reply:
Would that be okay keeping always an empty table though?

Reply:
We can purge the table as a last step. Let's say it purges files from the list which were removed 90 days ago. IDK what would happen if there was a file added that has the same content has and then also deleted. Would we have 2 entries in the table? If that's the case, and they don't clash, I don't see a problem with deleting these entries periodically. Deleting them straight away is also valid, you could get rid of the status field.

Reply:
Ok, I'll delete them directly then and will remove the status field.